### PR TITLE
feat(email): synthetic .mbox dataset for email triage tests

### LINKED ISSUE #848: feat(email): synthetic .mbox dataset for email triage agent testing
## Summary

The Email Triage Agent (Phase 2, v0.23.0) needs a synthetic `.mbox` dataset to validate the triage engine, MIME parsing, email threading, and categorization logic without requiring real email accounts or OAuth setup. This unblocks offline development and CI testing of the email pipeline described in the [email & calendar integration plan](https://github.com/amd/gaia/blob/main/docs/plans/email-calendar-integration.mdx).

## Why

- **No real inboxes in CI** — tests cannot depend on live IMAP/OAuth credentials
- **Reproducible triage evaluation** — ground-truth labels let us measure categorization accuracy against the [80% target](https://github.com/amd/gaia/blob/main/docs/plans/email-calendar-integration.mdx#12-success-metrics)
- **MIME edge cases** — multipart, attachments, inline images, non-UTF-8 encodings, and nested forwarding need coverage upfront
- **Threading correctness** — `In-Reply-To` / `References` chains must parse correctly for conversation grouping

## Proposed Dataset

Generate a synthetic `.mbox` file (Python `mailbox` + `email` stdlib) with **~200-250 messages** covering a realistic inbox mix.

### Triage categories (ground truth labels in metadata)

The plan (section 4.2) defines 4 triage categories. This dataset adds a **pre-triage spam/phishing class** — the triage engine should filter these *before* categorization, not treat them as a 5th category. Ground truth marks them with `is_spam: true` / `is_phishing: true` separately from `category`.

| Category | Count | Examples |
|----------|-------|---------|
| **Urgent** | ~20 | Boss escalation, client contract deadline, prod incident alert, security advisory |
| **Actionable** | ~45 | PR review requests, meeting invites, direct questions, expense approvals |
| **Informational** | ~55 | Newsletters, order receipts, shipping notifications, internal announcements |
| **Low priority** | ~30 | Marketing promos, social media notifications, cold outreach |
| **Spam/phishing** (pre-triage filter) | ~20 | Phishing attempts, fake invoices, scams, fake delivery notices |
| **Ambiguous/borderline** | ~15 | Hard-to-classify messages (see section below) |

Messages within each category are drawn from the corporate, personal, and spam flavors described below — the flavors are **content types within categories**, not separate pools.

### Recurring sender personas

Define 8-10 recurring senders so threading and sender-importance learning can be tested across multiple messages from the same person:

| Persona | Domain | Role | Typical category |
|---------|--------|------|-----------------|
| Sarah Chen (boss) | `@acme-corp.example.com` | VP Engineering | Urgent / Actionable |
| Alex Kumar (direct report) | `@acme-corp.example.com` | Senior Engineer | Actionable |
| Jordan Lee (cross-team) | `@acme-corp.example.com` | Product Manager | Actionable / Informational |
| IT Systems | `noreply@acme-corp.example.com` | Automated | Informational |
| HR Team | `hr@acme-corp.example.com` | Automated | Informational |
| Maria Santos (client) | `@globaltech.example.net` | External partner | Urgent / Actionable |
| DevOps Bot | `alerts@acme-corp.example.com` | CI/CD / PagerDuty | Urgent / Informational |
| Newsletter senders | various `@*.example.com` | Marketing / news | Informational / Low priority |

Each recurring sender should appear in **3-8 messages** (mix of thread roots and replies) to enable sender-frequency and response-pattern analysis.

### Corporate / enterprise email types (~60-70 messages)

The primary target is enterprise users on Outlook/Exchange. These messages are distributed across the triage categories above:

- **IT / Ops notifications** — password expiry warnings, VPN maintenance windows, system outage alerts, new software rollout announcements
- **HR / People Ops** — benefits enrollment reminders, PTO policy updates, org announcements, new hire intros, mandatory training deadlines
- **Executive comms** — all-hands meeting recap, quarterly earnings summary, CEO update, reorg announcement
- **Cross-team threads** — multi-person `Cc` chains with 4-6 participants, "loop in [name]" forwards, top-posted reply style
- **Calendar-related** — meeting invite (`.ics`), meeting cancellation, room booking confirmation, recurring meeting update
- **Compliance / Legal** — NDA reminder, export control notice, data retention policy update, audit request
- **Automated systems** — JIRA ticket assignment, CI/CD build failure, Confluence page update, PagerDuty alert, Salesforce lead notification
- **Expense / Finance** — expense report approval, purchase order confirmation, travel booking receipt, budget review request

### Spam / phishing (~20 messages, pre-triage filter)

Realistic spam that the triage engine should filter before categorization:

- **Phishing** — fake password reset, "verify your account", spoofed IT department, fake DocuSign
- **Scam** — Nigerian prince, lottery winner, inheritance notification, crypto "opportunity"
- **Commercial spam** — unsolicited product pitch, SEO services, fake invoice attachment, "limited time offer"
- **Social engineering** — fake LinkedIn connection, spoofed coworker name with external domain, urgent wire transfer request from "CEO"
- **Delivery scams** — fake UPS/FedEx tracking, "package held at customs", Amazon order you did not place

Spam messages should include realistic spam signals: mismatched `From` display name vs address, suspicious domains, urgency language, misspellings, suspicious attachment names, missing `List-Unsubscribe`, etc.

### Personal / consumer email types (~30 messages)

- Order confirmations and shipping updates (Amazon, retailer receipts)
- Subscription newsletters (tech blogs, news digests)
- Social media notifications (LinkedIn, GitHub stars/follows)
- Bank/financial alerts (transaction notification, statement ready)
- Travel confirmations (flight, hotel, car rental)
- App notifications (Slack digest, calendar reminders)

### Ambiguous / borderline messages (~15 messages)

Intentionally hard-to-classify emails that stress-test the triage engine decision boundaries. Mark these with `ambiguous: true` in ground truth, plus a `rationale` field explaining the intended classification:

- Meeting invite from unknown external contact (Actionable or Low priority?)
- Vendor invoice with no prior relationship (Urgent or Informational?)
- Automated JIRA ticket from a project the user is not on (Actionable or Informational?)
- Newsletter from a tool you actively use vs. one signed up for and forgotten (Informational or Low priority?)
- "Quick question" from someone never emailed before (Actionable or Low priority?)
- Internal compliance email that requires acknowledgment by EOD (Urgent or Actionable?)
- Reply-all on a thread where the user was Cc'd but not addressed (Actionable or Informational?)

### Malformed / edge-case messages (~5-10 messages)

Real inboxes contain broken email. Include parser-robustness tests:

- Missing `Subject` header
- Empty body (headers only)
- Truncated multipart (missing closing boundary)
- Invalid `Date` header (unparseable format)
- Double-encoded UTF-8 subject (`=?UTF-8?B?...?=` wrapping already-encoded text)
- Base64-encoded body with incorrect padding
- Extremely long `Subject` (>500 chars)
- Message with no `From` header

### Email structure variety

- Plain text only
- HTML only (with realistic corporate email templates)
- Multipart (text + HTML)
- With attachments (small dummy `.pdf`, `.csv`, `.png`, `.docx` — **1-5 KB each, total .mbox under 1 MB**)
- Inline images (`Content-Disposition: inline`)
- Forwarded messages (nested `message/rfc822`)
- Reply chains (3-5 deep with proper `In-Reply-To` / `References` headers)
- Non-ASCII subjects and bodies (UTF-8, ISO-8859-1)
- Calendar invites (`.ics` attachments)
- Top-posted replies with `>` quoted original (Outlook style)
- HTML signature blocks with logos and legal disclaimers

### Realistic metadata

- Varied `From` addresses using the recurring sender personas above
- Corporate domains: `@acme-corp.example.com`, `@globaltech.example.net` (RFC 2606 reserved)
- Realistic `Date` headers spanning ~2 weeks, **clustered around 9 AM and 4-5 PM** on weekdays to simulate real arrival patterns, with a batch arriving overnight Saturday to Monday for "overnight triage" testing
- Proper `Message-ID`, `In-Reply-To`, `References` for threading
- Mix of `To`, `Cc`, `Bcc` patterns (including large `Cc` lists for corporate threads)
- `List-Unsubscribe` headers on marketing/newsletter emails
- `X-Priority` / `Importance` headers on some urgent emails
- `X-Mailer` headers (Outlook, Thunderbird, Gmail, automated systems)
- Corporate email disclaimers in footers ("This email is confidential...")
- `Reply-To` mismatches on phishing emails
- `Received` header chains simulating realistic relay paths

## Deliverables

1. **Generator script** — `tests/fixtures/email/generate_mbox.py` that produces the `.mbox` deterministically (seeded RNG)
2. **Pre-built fixture** — `tests/fixtures/email/synthetic_inbox.mbox` checked into repo (must be under 1 MB)
3. **Ground truth manifest** — `tests/fixtures/email/ground_truth.json` mapping `Message-ID` to `{ category, priority, is_thread_root, thread_id, has_attachment, is_spam, is_phishing, ambiguous, rationale, sender_persona }`
4. **Pytest fixtures** — in `tests/fixtures/email/conftest.py`:
   - `synthetic_mbox` — loads the `.mbox` file
   - `email_ground_truth` — loads the ground truth JSON
   - `single_email(category)` — returns one message of a given triage category
   - `spam_emails` — returns all spam/phishing messages for filter testing
   - `ambiguous_emails` — returns borderline messages for boundary testing

## Acceptance criteria

- [ ] `python tests/fixtures/email/generate_mbox.py` produces a valid `.mbox` readable by Python `mailbox.mbox()`
- [ ] All 4 triage categories represented with ground truth labels, plus separate `is_spam`/`is_phishing` flags (spam is a pre-triage filter, not a 5th category)
- [ ] Corporate email types cover IT, HR, exec comms, cross-team threads, and automated systems
- [ ] Spam/phishing emails include realistic indicators (domain mismatches, urgency, suspicious attachments)
- [ ] ~15 ambiguous/borderline messages included with `ambiguous: true` and `rationale` in ground truth
- [ ] ~5-10 malformed messages test parser robustness (missing headers, truncated MIME, encoding errors)
- [ ] Recurring sender personas appear in 3-8 messages each for sender-importance testing
- [ ] Multipart, attachment, threading, and encoding edge cases covered
- [ ] No real PII — all names, addresses, domains, and content are synthetic (use `.example.com` / `.example.net`)
- [ ] Total `.mbox` file size under 1 MB; individual dummy attachments 1-5 KB each
- [ ] Pre-built fixture matches generator output (hash check in CI — `python tests/fixtures/email/generate_mbox.py --verify`)
- [ ] Pytest can load and iterate the dataset (`tests/unit/test_synthetic_mbox.py` passes)
- [ ] Generator is deterministic (same seed produces identical output)

## References

- Plan: [`docs/plans/email-calendar-integration.mdx`](https://github.com/amd/gaia/blob/main/docs/plans/email-calendar-integration.mdx) section 4 (Email Triage Agent)
- Related: #645 (Email Triage Agent), #663 (Daily briefs)

### PR DESCRIPTION
**Pull request by theonlychant**
Closes #848

   ## Summary
   Adds a synthetic .mbox dataset for testing the email triage agent.
   The fixtures provide realistic email threads for unit and integration
   testing without requiring live mailbox access.

   ## Why GAIA needs it
   The email triage agent currently has no test data to run against,
   making it impossible to validate triage logic in CI.

   ## Test plan
   - `tests/unit/test_agents_split.py` - all tests passing

### CODE REVIEW

File: src/gaia/agents/chat/agent.py [hunk 1]
```diff
@@ -1704,6 +1709,261 @@ def search_web(query: str) -> dict:
...
+                        r"document|pdf|index|search.*file|rag|chunk|summarize|summary",
+                        r"what does .+ say",
+                        r"find.*in .+(report|handbook|manual|doc)",
+                    }
+                ),
+            )
+        )
+
+        # Pre-activate RAG bundle when documents are already indexed
+<<<<<<< HEAD
```
Comment (itomek-amd) [inline src/gaia/agents/chat/agent.py]: Unresolved merge conflict committed here (lines 1775-1787). The nested `<<<<<<< HEAD` inside `<<<<<<< HEAD` suggests a bad re-merge attempt. Resolve by picking one of the two formattings of this `if` statement and removing all `<<<<<<<` / `=======` / `>>>>>>>` lines. Run `git diff --check` after to catch any others.

File: src/gaia/agents/chat/agent.py [hunk 2]
```diff
@@ -1704,6 +1709,261 @@ def search_web(query: str) -> dict:
...
+                ToolBundle(
+                    name=f"mcp_{server}",
+                    tools=frozenset(tools),
+                    policy=ActivationPolicy.SESSION,
+                )
+            )
+
+        # Log the bundle configuration
+        total = len(_TOOL_REGISTRY)
+<<<<<<< HEAD
```
Comment (itomek-amd) [inline src/gaia/agents/chat/agent.py]: Same issue here (lines 1947-1959): nested unresolved merge conflict. Both branches of the conflict produce the same `bundled = sum(...)` expression, just with different line wrapping — pick one and delete the markers.

File: tests/unit/test_pkce.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import importlib.util
+import os
+import re
+
+
+def _load_pkce_module():
+    path = os.path.join(os.path.dirname(__file__), "..", "..", "src", "gaia", "connections", "pkce.py")
```
Comment (itomek-amd) [inline tests/unit/test_pkce.py]: This test imports `src/gaia/connections/pkce.py`, but the last commit of this PR (`chore: remove connections/ — tracked separately in #915`) deleted that directory. The test will fail at collection time with `FileNotFoundError`. Since the corresponding source was removed, this whole file should be deleted too.

File: src/gaia/agents/chat/agent.py [hunk 4]
```diff
@@ -1704,6 +1709,261 @@ def search_web(query: str) -> dict:
...
+            and self.rag
+            and getattr(self.rag, "indexed_files", None)
+        ):
+=======
+        if hasattr(self, "rag") and self.rag and getattr(self.rag, "indexed_files", None):
+>>>>>>> 2f79fa97 (trying to address the issue in::Tool-registry scaling: resolve scratchpad/memory collision via #688 dynamic loading #800)
+=======
+        if hasattr(self, "rag") and self.rag and getattr(self.rag, "indexed_files", None):
+>>>>>>> 2f79fa97 (trying to address the issue in::Tool-registry scaling: resolve scratchpad/memory collision via #688 dynamic loading #800)
+            self.tool_loader._state["rag"].activated = True
```
Comment (itomek-amd) [inline src/gaia/agents/chat/agent.py]: (Non-blocking, for the follow-up tool-loader PR.) Reaching into `self.tool_loader._state["rag"].activated = True` couples ChatAgent to ToolLoader's private state. Worth a small public method like `ToolLoader.activate(bundle_name)` so callers don't depend on the internal `_state` dict's shape.

File: src/gaia/agents/base/tool_loader.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+    def get_active_bundle_names(self) -> list[str]:
+        """Return names of currently activated bundles."""
+        return [n for n, s in self._state.items() if s.activated]
+
+    def get_bundle_for_tool(self, tool_name: str) -> Optional[str]:
+        """Return the bundle name that owns *tool_name*, or ``None``."""
+        return self._tool_to_bundle.get(tool_name)
+
+    def reset_session(self) -> None:
```
Comment (itomek-amd) [inline src/gaia/agents/base/tool_loader.py:228]: (Non-blocking, for the follow-up tool-loader PR.) `reset_session()` is defined and tested but never called from the chat agent, so bundle activation persists for the entire process lifetime — across what users would consider distinct conversations. Probably wants to be wired into whatever marks "new conversation" in `ChatAgent`.

**Review by itomek-amd — CHANGES_REQUESTED** [review]
@theonlychant — thanks for the work, and the cleanup pass that pulled `connections/` back out was the right call. Requesting changes for two reasons:

1. **Code doesn't parse.** Two unresolved merge-conflict blocks landed in `src/gaia/agents/chat/agent.py` (see inline comments). `python -m py_compile src/gaia/agents/chat/agent.py` fails with `SyntaxError: invalid decimal literal at line 1784`, which is also why the `Claude AI Assistant / pr-review` check failed — the bot can't review a file that doesn't compile.
2. **Title and scope don't match.** The PR is titled `feat(email): synthetic .mbox dataset for email triage tests` and the body says `Closes #848`, but the diff bundles ~1100 lines of tool-loader work for #688/#800 and a `tests/unit/test_pkce.py` that imports a path (`src/gaia/connections/pkce.py`) the same PR deletes.

### Two ways to resolve

**Option A (recommended) — refocus this PR on its stated scope.** Revert the tool-loader changes (`src/gaia/agents/base/tool_loader.py`, the `tool_loader` wiring in `base/agent.py`, the new ~260 lines in `chat/agent.py`, and `tests/unit/test_tool_loader.py`). Delete `tests/unit/test_pkce.py`. What's left is a clean #848 PR. Open a separate PR for the tool-loader linked to #688/#800.

**Option B — keep both, but fix the merge.** Resolve the conflict markers in `chat/agent.py`, delete `tests/unit/test_pkce.py`, update the title and body to declare both threads (e.g. `feat: email .mbox fixtures + tool-loader for #688`), and link both issues. I'd push for A — bundled PRs are harder to land and harder to revert if one half regresses.

### Before re-requesting review

- `python -m py_compile src/gaia/agents/chat/agent.py` exits 0
- `pytest tests/unit/test_synthetic_mbox.py -q` is green on a clean checkout
- (If keeping any tool-loader work) `pytest tests/unit/test_tool_loader.py -q` is green

Happy to re-review once either option lands. The mbox fixture itself looks like a useful chunk of work.

**Comment by theonlychant** [comment]
Alright so @kovtcharov or @kovtcharov-amd and @itomek or @itomek-amd , I've addressed both issues:

Removed all tool-loader and pkce files and PR is now scoped to #848 only
Fixed the set_charset payload error in generate_mbox.py and all tests now passing

Ready for re-review when you get a chance, thanks!

File: src/gaia/agents/base/tool_loader.py (hunk 5 above)
Comment (theonlychant) [reply src/gaia/agents/base/tool_loader.py:228]: alright I will look into this

Comment (theonlychant) [reply src/gaia/agents/base/tool_loader.py:228]: check the changes at [commit](https://github.com/amd/gaia/pull/928/changes/04e0ad7c21537b002ae7b9743d1c4cec2f970b9d) 

```bash
:~/gaia$  PYTHONPATH=src python3 -m pytest tests/test_chat_agent.py -q
............................................                             [100%]
=============================== warnings summary ===============================
../.local/lib/python3.13/site-packages/_pytest/config/__init__.py:1434
  /home/theonlychant/.local/lib/python3.13/site-packages/_pytest/config/__init__.py:1434: PytestConfigWarning: Unknown config option: asyncio_mode
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type swigvarlink has no __module__ attribute

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
44 passed, 4 warnings in 23.36s
```
the tests were ok

**Review by itomek — APPROVED** [review]

**Review by itomek-amd — APPROVED** [review]
